Source/SwiftLintFramework/Rules/Idiomatic/NoConstantConditionRule.swift

```diff
@@ -0,0 +1,566 @@
+            Example("guard ↓2 < 3 {} else {}"),
+            Example("switch ↓2 < 3 { default: break }"),
+            Example("switch ↓\"foo\" { default: break }"),
+            Example("switch ↓foo() && false { default: break }")
+        ]
+    )
+
+    func preprocess(file: SwiftLintFile) -> SourceFileSyntax? {
+        return file.foldedSyntaxTree
+    }
+
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        return NoConstantConditionViolationsVisitor(viewMode: .fixedUp)
+    }
+}
+
+private enum ConstantValue: Equatable {
+    case booleanValue(_ value: Bool)
+    case integerValue(_ value: Int)
+    case floatValue(_ value: Float)
+    case stringValue(_ value: String)
+    case nilValue
+}
+
+private enum ConstantEvaluationResult: Equatable {
+    case evaluated(_ value: ConstantValue)
+    case notEvaluated
+}
+
+private enum ExpressionResult: Equatable {
+    case alwaysEvaluatesToFalse
+    case evaluatedConstant(_ value: ConstantValue)
+    case nonEvaluatedButConstant
+    case dynamic
+}
+
+private func getTokenText(_ token: TokenSyntax) -> String {
+    return token.text
+}
+
+private enum StringLiteralResult {
+    case constant(_ value: String)
+    case dynamic
+}
+
+private func analyzeStringLiteral(_ expression: StringLiteralExprSyntax) -> StringLiteralResult {
+    var value = ""
+
+    for literalSegment in expression.segments {
+        switch literalSegment {
+        case .expressionSegment:
+            return .dynamic
+        case let .stringSegment(stringSegment):
+            value += getTokenText(stringSegment.content)
+        }
+    }
+
+    return .constant(value)
+}
+
+private func analyzePrimitiveLiteralExpression(_ element: ExprSyntax) -> ExpressionResult {
+    if let booleanExpression = element.as(BooleanLiteralExprSyntax.self) {
+        let value = getTokenText(booleanExpression.booleanLiteral)
+        return .evaluatedConstant(.booleanValue(value == "true"))
+    }
+
+    if let integerExpression = element.as(IntegerLiteralExprSyntax.self) {
+        let value = getTokenText(integerExpression.digits)
+        return .evaluatedConstant(.integerValue(Int(value) ?? 0))
+    }
+
+    if let floatExpression = element.as(FloatLiteralExprSyntax.self) {
+        let value = getTokenText(floatExpression.floatingDigits)
+        return .evaluatedConstant(.floatValue(Float(value) ?? 0.0))
+    }
+
+    if let stringExpression = element.as(StringLiteralExprSyntax.self) {
+        let stringLiteralResult = analyzeStringLiteral(stringExpression)
+        switch stringLiteralResult {
+        case let .constant(value):
+            return .evaluatedConstant(.stringValue(value))
+        case .dynamic:
+            return .dynamic
+        }
+    }
+
+    if element.is(NilLiteralExprSyntax.self) {
+        return .evaluatedConstant(.nilValue)
+    }
+
+    return .dynamic
+}
+
+private func isBinaryOperator(_ operatorExpression: ExprSyntax, _ expectedOperator: String) -> Bool {
```

Comment:
If you look at the structure of some of the other rules, in cases with a pure function like this, operating on an `ExprSyntax` (or lots of the other common objects), they'll often go for

```
private extension ExprSyntax {
    func isBinaryOperator(_ expectedOperator: String) -> Bool {
        // some code here
    }
}
```

Just search for `extension ExprSyntax` to see some examples.

I think a lot of code here *could* be given a similar treatment.

I don't think that's really documented anywhere, or mandatory in anyway, but just an observation that a lot of the existing code seems to follow that pattern.